build_tools/cmake/iree_static_linker_test.cmake

```diff
@@ -0,0 +1,259 @@
+      DEPENDS ${_COMPILER_TOOL} ${_RULE_SRC}
+    )
+
+    set(_EMITC_LIB_NAME "${_NAME}_emitc")
+    add_library(${_EMITC_LIB_NAME}
+      STATIC
+      ${_C_FILE_NAME}
+    )
+    target_compile_definitions(${_EMITC_LIB_NAME} PUBLIC EMITC_IMPLEMENTATION)
+    SET_TARGET_PROPERTIES(
+      ${_EMITC_LIB_NAME}
+      PROPERTIES
+        LINKER_LANGUAGE C
+    )
+  else()  # bytecode module path
+    set(_VMFB_FILE_NAME ${_RULE_NAME}.vmfb)
+    list(APPEND _COMPILER_ARGS "-o")
+    list(APPEND _COMPILER_ARGS "${_VMFB_FILE_NAME}")
+
+    # Custom command for iree-compile to generate static library and VM module.
+    add_custom_command(
+      OUTPUT
+        ${_H_FILE_NAME}
+        ${_O_FILE_NAME}
+        ${_VMFB_FILE_NAME}
+      COMMAND ${_COMPILER_TOOL} ${_COMPILER_ARGS}
+      DEPENDS ${_COMPILER_TOOL} ${_RULE_SRC}
+    )
+
+    # Generate the embed data with the bytecode module.
+    set(_MODULE_C_NAME "${_RULE_NAME}_module_c")
+    set(_EMBED_H_FILE_NAME ${_MODULE_C_NAME}.h)
+    set(_EMBED_C_FILE_NAME ${_MODULE_C_NAME}.c)
+    iree_c_embed_data(
+      NAME
+        "${_MODULE_C_NAME}"
+      IDENTIFIER
+        "${_NAME}"
+      GENERATED_SRCS
+        "${_VMFB_FILE_NAME}"
+      C_FILE_OUTPUT
+        "${_EMBED_C_FILE_NAME}"
+      H_FILE_OUTPUT
+        "${_EMBED_H_FILE_NAME}"
+      FLATTEN
+      PUBLIC
+    )
+  endif(_RULE_EMITC)
+
+  set(_LIB_NAME "${_NAME}_lib")
+  add_library(${_LIB_NAME}
+    STATIC
+    ${_O_FILE_NAME}
+  )
+  SET_TARGET_PROPERTIES(
+    ${_LIB_NAME}
+    PROPERTIES
+    LINKER_LANGUAGE C
+  )
+  # Set alias for this static library to be used later in the function.
+  add_library(${_PACKAGE_NS}::${_RULE_NAME}_lib ALIAS ${_LIB_NAME})
+
+  if(NOT _RULE_INPUT_SHAPES OR
+     NOT _RULE_INPUT_TYPE OR
+     NOT _RULE_HAL_TYPE OR
+     NOT _RULE_MAIN_FUNCTION OR
+     NOT _RULE_STATIC_LIB_PREFIX)
+    return()
+  endif()
+
+  # Process module input configs
+  list(LENGTH _RULE_INPUT_SHAPES _INPUT_NUM)
+
+  set(_INPUT_DIM_LIST)
+  set(_INPUT_SIZE_LIST)
+  set(_INPUT_SHAPE_STR "\{")
+  set(_INPUT_DIM_MAX 0)
+  foreach(_INPUT_SHAPE_ENTRY ${_RULE_INPUT_SHAPES})
+    set(_INPUT_SHAPE_STR "${_INPUT_SHAPE_STR}\{${_INPUT_SHAPE_ENTRY}\}, ")
+
+    string(REPLACE "," ";" _INPUT_SHAPE_LIST "${_INPUT_SHAPE_ENTRY}")
+    list(LENGTH _INPUT_SHAPE_LIST _INPUT_DIM)
+    list(APPEND _INPUT_DIM_LIST ${_INPUT_DIM})
+    if(_INPUT_DIM GREATER _INPUT_DIM_MAX)
+      set(_INPUT_DIM_MAX ${_INPUT_DIM})
+    endif()
+    set(_INPUT_SIZE 1)
+    foreach(_INPUT_DIM_VAL ${_INPUT_SHAPE_LIST})
+      math(EXPR _INPUT_SIZE "${_INPUT_SIZE} * ${_INPUT_DIM_VAL}")
+    endforeach()
+    list(APPEND _INPUT_SIZE_LIST ${_INPUT_SIZE})
+  endforeach()
+  set(_INPUT_SHAPE_STR "${_INPUT_SHAPE_STR}\}")
+  string(REPLACE ";" ", " _INPUT_DIM_STR "${_INPUT_DIM_LIST}")
+  string(REPLACE ";" ", " _INPUT_SIZE_STR "${_INPUT_SIZE_LIST}")
+
+  # Generate the source file.
+  # TODO(scotttodd): Move to build time instead of configure time?
```

Comment:
Not sure if we need to keep this TODO. Put it here in case we want to migrate template from `configure_file` to other mechanisms.

Reply:
I don't see what the problem is having this at configure time

Reply:
When I was writing the HAL CTS suite, I didn't like how many source files would be generated at configure time. I want configure to be faster, not slower (though I don't think it actually adds _that much_ time)

Reply:
The other problem is that these .c files will not be cleaned by `--target clean`, and if you accidentally remove them, they will not be re-generated until you trigger cmake reconfigure.

One possibility is to use other template tools (e.g., python template) so we can use `add_custom_command` to generate the file with the right dependency, but that's a different LSC.